functions/go/gatekeeper-validate/validate.go

```diff
@@ -0,0 +1,171 @@
+	opatypes "github.com/open-policy-agent/frameworks/constraint/pkg/types"
+	"github.com/open-policy-agent/gatekeeper/pkg/target"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+var scheme = runtime.NewScheme()
+
+func init() {
+	err := v1beta1.SchemeBuilder.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func createClient() (*opaclient.Client, error) {
+	driver := local.New(local.Tracing(false))
+	backend, err := opaclient.NewBackend(opaclient.Driver(driver))
+	if err != nil {
+		return nil, err
+	}
+	return backend.NewClient(opaclient.Targets(&target.K8sValidationTarget{}))
+}
+
+func gatherTemplates(objects []runtime.Object) ([]*templates.ConstraintTemplate, error) {
+	var templs []*templates.ConstraintTemplate
+	for _, obj := range objects {
+		ct, isConstraintTemplate := obj.(*v1beta1.ConstraintTemplate)
+		if !isConstraintTemplate {
+			continue
+		}
+		templ := &templates.ConstraintTemplate{}
+		if err := scheme.Convert(ct, templ, nil); err != nil {
+			return nil, err
+		}
+		templs = append(templs, templ)
+	}
+	return templs, nil
+}
+
+func gatherConstraints(objects []runtime.Object) ([]*unstructured.Unstructured, error) {
+	var cstrs []*unstructured.Unstructured
+	for _, obj := range objects {
+		gvk := obj.GetObjectKind().GroupVersionKind()
+		if gvk.Group == "constraints.gatekeeper.sh" {
+			cstrs = append(cstrs, obj.(*unstructured.Unstructured))
+		}
+	}
+	return cstrs, nil
+}
+
+// Validate makes sure the configs passed to it comply with any Constraints and
+// Constraint Templates present in the list of configs
+func Validate(objects []runtime.Object) error {
+	client, err := createClient()
+	if err != nil {
+		return err
+	}
+	tmpls, err := gatherTemplates(objects)
+	if err != nil {
+		return err
+	}
+	ctx := context.Background()
+	for _, t := range tmpls {
+		if _, err = client.AddTemplate(ctx, t); err != nil {
+			return err
+		}
+	}
+	cstrs, err := gatherConstraints(objects)
+	if err != nil {
+		return err
+	}
+	for _, c := range cstrs {
+		if _, err = client.AddConstraint(ctx, c); err != nil {
+			return err
+		}
+	}
+
+	for _, obj := range objects {
+		if _, err = client.AddData(ctx, obj); err != nil {
+			return err
+		}
+	}
+
+	resps, err := client.Audit(ctx)
+	if err != nil {
+		return err
+	}
+	results := resps.Results()
+	if len(results) > 0 {
+		return parseResults(results)
+	}
+	return nil
+}
+
+func parseResults(results []*opatypes.Result) error {
+	out := &framework.Result{
+		Items: []framework.Item{},
+	}
+
+	for _, r := range results {
+		u, ok := r.Resource.(*unstructured.Unstructured)
+		if !ok {
+			return fmt.Errorf("could not cast to unstructured: %+v", r.Resource)
+		}
+
+		item := framework.Item{
+			Message:  fmt.Sprintf("%s\nviolatedConstraint: %s", r.Msg, r.Constraint.GetName()),
+			Severity: framework.Error,
```

Comment:
Gatekeeper has a few different possible violation levels for a constraint now:

- `deny`
- `warn`
- `dryrun`

I'm not sure how you'll want to interpret these, (e.g. for the admission webhook,  `warn` allows the request but sends a warning ), but that could be useful metadata for users.

Reply:
Updated.
PTAL